deepchem/hyper/gaussian_process.py

```diff
@@ -0,0 +1,185 @@
+  """
+
+  def hyperparam_search(self,
+                        params_dict,
+                        train_dataset,
+                        valid_dataset,
+                        output_transformers,
+                        metric,
+                        n_features=1024,
+                        n_tasks=1,
+                        max_iter=20,
+                        search_range=4,
+                        hp_invalid_list=[
+                            'seed', 'nb_epoch', 'penalty_type', 'dropouts',
+                            'bypass_dropouts', 'n_pair_feat'
+                        ],
+                        logdir=None):
+    """Perform hyperparams search using a gaussian process assumption
+
+    params_dict include single-valued parameters being optimized,
+    which should only contain int, float and list of int(float)
+
+    parameters with names in hp_invalid_list will not be changed.
+
+    For Molnet models, self.model_class is model name in string,
+    params_dict = dc.molnet.preset_hyper_parameters.hps[self.model_class]
+    """
+
+    assert len(metric) == 1, 'Only use one metric'
+    hyper_parameters = params_dict
+    hp_list = hyper_parameters.keys()
+    for hp in hp_invalid_list:
+      if hp in hp_list:
+        hp_list.remove(hp)
+
+    hp_list_class = [hyper_parameters[hp].__class__ for hp in hp_list]
+    assert set(hp_list_class) <= set([list, int, float])
+    # Float or int hyper parameters(ex. batch_size, learning_rate)
+    hp_list1 = [
+        hp_list[i] for i in range(len(hp_list)) if not hp_list_class[i] is list
+    ]
+    # List of float or int hyper parameters(ex. layer_sizes)
+    hp_list2 = [(hp_list[i], len(hyper_parameters[hp_list[i]]))
+                for i in range(len(hp_list)) if hp_list_class[i] is list]
+
+    # Number of parameters
+    n_param = len(hp_list1) + sum([hp[1] for hp in hp_list2])
+    # Range of optimization
+    param_range = []
+    for hp in hp_list1:
+      if hyper_parameters[hp].__class__ is int:
+        param_range.append((('int'), [
+            hyper_parameters[hp] / search_range,
+            hyper_parameters[hp] * search_range
+        ]))
+      else:
+        param_range.append((('cont'), [
+            hyper_parameters[hp] / search_range,
+            hyper_parameters[hp] * search_range
+        ]))
+    for hp in hp_list2:
+      if hyper_parameters[hp[0]][0].__class__ is int:
+        param_range.extend([(('int'), [
+            hyper_parameters[hp[0]][i] / search_range,
+            hyper_parameters[hp[0]][i] * search_range
+        ]) for i in range(hp[1])])
+      else:
+        param_range.extend([(('cont'), [
+            hyper_parameters[hp[0]][i] / search_range,
+            hyper_parameters[hp[0]][i] * search_range
+        ]) for i in range(hp[1])])
+
+    # Dummy names
+    param_name = ['l' + format(i, '02d') for i in range(20)]
+    param = dict(zip(param_name[:n_param], param_range))
+
+    def f(l00=0,

```

Comment:
What's the purpose of this helper function? Could you add a comment explaining what it does?